mmv1/third_party/terraform/functions/name_from_id_test.go

```diff
@@ -0,0 +1,88 @@
+package functions_test
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-provider-google/google/acctest"
+)
+
+func TestAccProviderFunction_name_from_id(t *testing.T) {
+	t.Parallel()
+	// Skipping due to requiring TF 1.8.0 in VCR systems : https://github.com/hashicorp/terraform-provider-google/issues/17451
+	acctest.SkipIfVcr(t)
+
+	context := map[string]interface{}{
+		"function_name": "name_from_id",
+		"output_name":   "name",
+		"resource_name": fmt.Sprintf("tf-test-name-id-func-%s", acctest.RandString(t, 10)),
+	}
+
+	nameRegex := regexp.MustCompile(fmt.Sprintf("^%s$", context["resource_name"]))
+
+	acctest.VcrTest(t, resource.TestCase{
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				// Can get the project from a resource's id in one step
+				// Uses google_pubsub_topic resource's id attribute with format projects/{{project}}/topics/{{name}}
+				Config: testProviderFunction_get_project_from_resource_id(context),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestMatchOutput(context["output_name"].(string), nameRegex),
+				),
+			},
+			{
+				// Can get the project from a resource's self_link in one step
+				// Uses google_compute_disk resource's self_link attribute
+				Config: testProviderFunction_get_project_from_resource_self_link(context),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestMatchOutput(context["output_name"].(string), nameRegex),
+				),
+			},
+		},
+	})
+}
+
+func testProviderFunction_get_name_from_resource_id(context map[string]interface{}) string {
+	return acctest.Nprintf(`
+# terraform block required for provider function to be found
+terraform {
+  required_providers {
+    google = {
+      source = "hashicorp/google"
+    }
+  }
+}
+
+resource "google_pubsub_topic" "default" {
+  name = "%{resource_name}"
+}
+
+output "%{output_name}" {
+  value = provider::google::%{function_name}(google_pubsub_topic.default.id)
```

Comment:
Provider functions can't imply the provider like resources do? The provider block will always be needed?

Reply:
That's right, the `provider::google::` syntax cannot be skipped and [it's not recommended to include provider names in function names](https://developer.hashicorp.com/terraform/plugin/best-practices/naming#function-names:~:text=Unlike%20resource%20and%20data%20source%20naming%2C%20function%20names%20should%20NOT%20include%20the%20provider%20name%20since%20the%20configuration%20language%20syntax%20for%20calling%20the%20function%20must%20already%20specify%20the%20provider%20name%20separately).